Add JSON schema

franzpoeschel opened the pull request:
TODO:

- [x] Define particlePatches
- [x] Fill in things like "description", "type", "title", ...
- [x] ~~Maybe add support for abbreviated JSON representations~~ Not now
- [x] Dataset layout

I have added a CLI tool `openpmd-convert_json_toml`, exposing internally existing functionality to convert between TOML and JSON. I think it's better to define and maintain the schema in terms of TOML (it can be commented and is more legible), and just compile it to JSON for verification.

x-ref #1436

franzpoeschel committed:
Add JSON schema for openPMD

Written as .toml files for ease of documentation, maintailability and
readability.

franzpoeschel committed:
Add convert-json-toml tool

Needed for "compiling" the schema to JSON
Also add a Makefile to further simplify this

franzpoeschel committed:
Add script for checking openPMD file against the schema

Workflow documented in README.md

franzpoeschel committed:
Don't use spaces in SerialIOTest attribute names

The JSON schema verification package does not like that

franzpoeschel committed:
Fix bugs detected by this verifier

Both of the form "data not found in places where data was expected"

franzpoeschel committed:
Add GitHub workflow

Verify all JSON-openPMD files written by testing against the schema

franzpoeschel committed:
Shorthand attributes

franzpoeschel committed:
Add dataset template mode

ax3l commented on .github/workflows/linux.yml:
```diff
@@ -253,6 +253,22 @@ jobs:
         cmake --build build --parallel 2
         ctest --test-dir build --output-on-failure
 
+        python3 -m pip install jsonschema
```
Extra section might be cleaner:
```suggestion
    - name: Schema Check
      run: |
        python3 -m pip install jsonschema
```

ax3l commented on .github/workflows/linux.yml:
```diff
@@ -253,6 +253,22 @@ jobs:
+        python3 -m pip install jsonschema
+        cd share/openPMD/json_schema
+        PATH="../../../build/bin:$PATH" make -j 2
+        # We need to exclude the thetaMode example since that has a different
+        # meshesPath and the JSON schema needs to hardcode that.
```
Are we able to patch this in `check.py`?

franzpoeschel replied:
Not very easily. The JSON schema is on the file system and the single .json files refer to each other by their file names. Changing this would require (1) traversing the entire JSON schema and overriding the meshes path, the particles path and the references and (2) somehow setting up python-jsonschema to cross-reference in-memory schemas which I don't even know if it supports that, both at runtime of `check.py`.

ax3l reviewed (commented):
Great work, thank you!

Wondering a bit if validation of an XML backend would be easier.... not sure.

franzpoeschel commented:
> Wondering a bit if validation of an XML backend would be easier.... not sure.

The major issue of the JSON schema approach is that it is not a predictive parser (https://en.wikipedia.org/wiki/Recursive_descent_parser), meaning that it deals with `{"anyOf": [<option1>, <option2>, ...]}`-like statements very naively. If there is an error somewhere deep down the hierarchy, the JSON schema just sees that all options have failed and cannot give any more details where the failure was. 

If we want to seriously consider XML as an alternative, my opinion is that it must at least solve this particular issue .. which I doubt it can properly, since this involves a bit of parser theory and grammar transformation.

This job is better done by handwritten parsers, which the openPMD-api and the openPMD-validator are.

However, if we add an XML backend at some point, an XML schema is definitely at least worth a consideration.

franzpoeschel committed:
Fix path

franzpoeschel committed:
Fix reading from stdin

franzpoeschel committed:
toml11 4.0 compatibility

franzpoeschel committed:
Only check for existing Iterations in writeOnly mode

franzpoeschel committed:
Some additions to schema

1. Support UNDEFINED datasets in template mode
2. gridUnitSI may now be a vector

franzpoeschel committed:
Remove deprecated jsonschema.validators.RefResolver

Apparently it's better to make everything 100 times more complicated

franzpoeschel committed:
Use most recent version of jsonschema

franzpoeschel committed:
Allow empty variable-based series

franzpoeschel committed:
Use if-then-else for better-steered parsing

anyOf and oneOf now only used for trivial distinctions, this makes
schemas much more robust since errors can be caught early and error
messages become actually useful.

franzpoeschel committed:
hmm

franzpoeschel committed:
Remove json cfg after test

Otherwise CI thinks this is an openPMD file

ax3l commented on share/openPMD/json_schema/README.md:
````diff
@@ -0,0 +1,47 @@
+openpmd_convert-json-toml @series.toml > series.json
+```
+
+A `Makefile` is provided in this folder to simplify the application of this conversion tool.
````
Makefile? Isn't this just built with CMake as one of the CLI tools?

ax3l replied:
Oh, I see. The makefile does the conversion of the schema toml files

ax3l replied:
```suggestion
A `Makefile` is provided in this folder to simplify the application of this conversion tool to the `.toml` files in this folder.
```

franzpoeschel replied:
Maybe a bit clearer: A `Makefile` is provided in this folder to automate generating the needed JSON files from the TOML files.
?

ax3l commented on CMakeLists.txt:
```diff
@@ -685,11 +685,12 @@ set(openPMD_TEST_NAMES
     pipe
 )
-set(openPMD_PYTHON_CLI_MODULE_NAMES ${openPMD_CLI_TOOL_NAMES})
+set(openPMD_PYTHON_CLI_MODULE_NAMES ls)
```
This line change looks like a hack?

franzpoeschel replied:
Not really. Until now, both `openPMD_CLI_TOOL_NAMES` and `openPMD_PYTHON_CLI_MODULE_NAMES` were identical since both contained only `openpmd-ls`. 
But they're not identical in general; and now, we additionally have `openpmd-convert-json-toml` as a CLI tool, but it is not written in Python.

ax3l commented on CMakeLists.txt:
```diff
@@ -685,11 +685,12 @@ set(openPMD_TEST_NAMES
 # command line tools
 set(openPMD_CLI_TOOL_NAMES
     ls
+    convert-json-toml
```
Also add this to `setup.py` -> `entry_points` -> `console_scripts`?

franzpoeschel replied:
This is not a Python module, but written in C++, so I doubt that will work?

ax3l commented on share/openPMD/json_schema/Makefile:
```diff
@@ -0,0 +1,15 @@
+convert := openpmd-convert-json-toml
```
Naming it `openpmd-convert-toml-json` seems to be more intuitive for what we use it here, but you mentioned anyway that it works bi-directional?

franzpoeschel replied:
It is bi-directional, but we can flip the name, since that's the intended use here.

ax3l commented on share/openPMD/json_schema/README.md:
```diff
@@ -0,0 +1,47 @@
+
+The openPMD standard is not entirely expressible in terms of a JSON schema:
+
+* Many semantic dependencies, e.g. that the `position/x` and `position/y` vector of a particle species be of the same size, or that the `axisLabels` have the same dimensionality as the dataset itself, will go unchecked.
```
```suggestion
* Many semantic dependencies, e.g., that the `position/x` and `position/y` vectors of a particle species need to be of the same size, or that the `axisLabels` have the same dimensionality as the dataset itself, will go unchecked.
```

ax3l commented on share/openPMD/json_schema/README.md:
```diff
@@ -0,0 +1,47 @@
+* Many semantic dependencies, e.g. that the `position/x` and `position/y` vector of a particle species be of the same size, or that the `axisLabels` have the same dimensionality as the dataset itself, will go unchecked.
+* The `meshesPath` is assumed to be `meshes/` and the `particlesPath` is assumed to be `particles/`. This dependency cannot be expressed.
+
+While a large part of the openPMD standard can indeed be verified by checking against a JSON schema, the standard is generally large enough to make this approach come to its limits. Verification of a JSON schema is similar to the use of a naive recursive-descent parser. Error messages will often be unexpectedly verbose and not very informative.
```
Might clarify a bit:
```suggestion
While a large part of the openPMD standard can indeed be verified by checking against a static JSON schema, the standard is generally large enough to make this approach come to its limits. Verification of a JSON schema is similar to the use of a naive recursive-descent parser. Error messages will often be unexpectedly verbose and not very informative.
```

ax3l commented on share/openPMD/json_schema/README.md:
````diff
@@ -0,0 +1,47 @@
+}
+```
+
+Support for an abbreviated notation such as `"meshesPath": "meshes/"` is currently not (yet) available.
````
In this section, you might want to advertise & link openPMD-validator again.

franzpoeschel replied:
Will do. We might additionally advertise the JSON schema in the validator. Adding the JSON validation to our CI brought up a number of bugs after all.

franzpoeschel committed:
Update documentation, rename convert-toml-json tool

franzpoeschel committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

franzpoeschel committed:
Add reference to openPMD-validator

franzpoeschel committed:
Update README.md

ax3l reviewed (approved):
LGTM :+1:
